python/mpt/src/mpt/__init__.py

```diff
@@ -0,0 +1,1372 @@
+        for address, account_diff in state_diff.account_diffs.items():
+            # First: Process storage updates
+            for key, value in account_diff.storage_updates.items():
+                if value == U256(0):
+                    self.delete_storage_key(address, key)
+                else:
+                    self.upsert_storage_key(address, key, rlp.encode(value))
+
+            # Second: Process account update or deletion
+            if account_diff.account is None:
+                self.delete_account(address)
+            else:
+                # Store code
+                self.codes[keccak256(account_diff.account.code)] = (
+                    account_diff.account.code
+                )
+
+                # Get current storage root (after storage updates)
+                storage_root = self.get_storage_root(address)
+
+                # Update account with correct storage root
+                encoded_account = encode_account(account_diff.account, storage_root)
+                self.upsert(keccak256(address), encoded_account)
+
+
+def decode_node(node: Bytes) -> InternalNode:
+    """Decode an RLP encoded node into an InternalNode."""
+
+    decoded = rlp.decode(node)
+
+    if isinstance(decoded, list) and len(decoded) == 17:
+        logger.debug("Decoding as branch node")
+        return BranchNode(subnodes=tuple(decoded[0:16]), value=decoded[16])
+    elif isinstance(decoded, list) and len(decoded) == 2:
+        logger.debug("Decoding as extension or leaf node")
+        prefix = decoded[0]
+        value = decoded[1]
+
+        if not isinstance(prefix, bytes):
+            raise ValueError(f"Invalid prefix type: {type(prefix)}")
+
+        # Determine if it's a leaf or extension node based on first nibble
+        first_nibble = prefix[0] >> 4
+        is_leaf = first_nibble in (2, 3)
+        logger.debug(f"First nibble: {first_nibble}, is_leaf: {is_leaf}")
+
+        # Extract the path from the compact encoding
+        nibbles = []
+        for b in prefix:
+            nibbles.extend([(b >> 4) & 0xF, b & 0xF])
+
+        # Remove the flag nibble and odd padding if present
+        if first_nibble in (1, 3):  # odd length
+            nibbles = nibbles[1:]
+            logger.debug("Odd length, removed first nibble")
+        else:  # even length
+            nibbles = nibbles[2:]
+            logger.debug("Even length, removed first two nibbles")
+
+        current_path = bytes(nibbles)
+        logger.debug(f"Current path: {nibble_path_to_hex(current_path)}")
+
+        if is_leaf:
+            return LeafNode(rest_of_key=current_path, value=value)
+        else:
+            return ExtensionNode(key_segment=current_path, subnode=value)
+    else:
+        raise ValueError(f"Unknown node structure: {type(decoded)}")
+
+
+# Redefinition of encode_internal_node from ethereum.cancun.trie
+# without keccak256 of the RLP encoded node if its length is greater than 32 bytes
```

Comment:
what's the reason? to reduce # of keccak done?

Reply:
no, it's because in encode_internal_node from ethereum lib, it automatically hashes the result of the RLP encoding if its length is >= 32, but sometimes in our implementation I thought we might not want to hash in some instances. 

after checking all the times i called this function, we always hash. i'll replace this by the real function

Reply:
ERRATUM: i need to separate the two steps -> RLPencode & hash, to store encoded nodes at a specific hash
EELS's `encode_internal_node` does not let me do that